### ISSUE TITLE
Reading from multiple changefeeds in Python

### ISSUE THREAD

**Issue by AtnNn** [rethinkdb/rethinkdb#3298]
The `next` operation in Python on cursors is a blocking operation, making it impossible to read from multiple changefeeds at once. This makes it very hard to take full advantage of changefeeds in Python.

**Comment by gchpaco** [comment]
So I was looking around for this and I found the following, which is essentially a backport of the Tulip project to Python 2 & earlier Python 3s.  Tulip is the reference implementation of what became `asyncio` in Python 3.4.  http://trollius.readthedocs.org/en/latest/index.html

**Comment by gchpaco** [comment]
Ruby equivalent is #3678.

**Comment by danielmewes** [comment]
Reading through this again, I think @coffeemug's suggestion https://github.com/rethinkdb/rethinkdb/issues/3298#issuecomment-62483313 is very nice interface as a baseline.
One thing I especially like about it is that we (and third-party driver developers) could pretty easily provide this interface across a range of languages. We could also support it in the Ruby driver at least as far as I can tell.

In addition to this interface, we can think about providing *additional* language-specific interfaces, such as one using e.g. Tulip or Python 3.4 asyncio.

**Comment by danielmewes** [comment]
If we follow https://github.com/rethinkdb/rethinkdb/issues/3298#issuecomment-62483313, should we allow `on(cb)` on all queries or only cursors?
```
q1 = r.table('foo').run(conn).on(fn1)
q2 = r.table('bar').get('k').run(conn).on(fn2)
conn.poll(wait=True)
```

**Comment by gchpaco** [comment]
I can live with `.on()` as a baseline, but I think we will want to do more specific things where natural (e.g. Python, Ruby).

**Comment by deontologician** [comment]
I think asyncio, while imperfect, is a good idea for python 3. But it heavily relies on `yield from` which is a syntactic construct that's not efficiently portable to python 2.

My suggestion is to use [gevent](http://www.gevent.org/). It's python2/3 compatible and allows the user to monkey patch the socket interface to interact with green threads / tasklets.

**Comment by deontologician** [comment]
Personally, I think callbacks are fairly non-pythonic, though this is definitely a matter of taste

**Comment by mlucy** [comment]
If we do use `on`, I think it should work for everything, not just changefeeds.

**Comment by coffeemug** [comment]
The nice thing about the `on`/`poll` interface is that it doesn't enforce a threading model and can be used as a low level API to implement pretty much any async interface anyone wants. But it is low level, and I don't think it's particularly usable by a general audience.

Since feeds are going to be getting all the spotlight, I think we should definitely do a better job of integrating into native language ecosystems (which, admittedly, is a pain in the neck).

**Comment by danielmewes** [comment]
I don't understand how you would use `on/poll` to integrate it with another async interface (e.g. some existing event queue). I think you'll always have to modify the driver to do so.

That being said, it still serves (at least) two purposes:
- It will make sure that our drivers are capable of handling concurrent queries, and we can test that in a consistent way across languages.
- We provide a fallback interface for users a) who don't want to or cannot rewrite their apps to use an async library, or b) who are using an async solution which we don't support yet.

Also writing drivers for all the important async libraries will probably take us some time, and this will provide an intermediate solution until we get there.

**Comment by larkost** [comment]
I don't like the `.on` interface as a solution to this problem. It is just moving the blocking problem to a different location and imposing its own async style. For example, I can't do the following without going into threads:

1. start a changefeed
2. check if there is data available
3. do something in the case that there is no data available
4. loop back to 2

The `get_nowait` solution that I talked about above does solve this problem, without trying to shoehorn an async model into this problem (which I think is a separate issue). Since I suggested this I actually have re-thought the command, and think we should just modify our existing `.next()` command to add an optarg such as `wait` that defaults to `True`, but can also take `False` or a float value (seconds to wait for something to be available). The rest of the suggestion still stands: if nothing arrives raise a subclass of `Queue.Empty`.

**Comment by danielmewes** [comment]
@larkost What I don't understand about the `get_nowait` model is how you would be waiting on multiple changefeeds without busy waiting, which isn't really a good option.

Maybe we can do a combination of your suggestion and the `on`/`poll` interface. Namely we could keep the `poll`, but use a non-blocking `next` (or `get_nowait()`) instead of callbacks / `on`:
```Python
while True
  feedA = r.table('a').changes().run(conn)
  feedB = r.table('b').changes().run(conn)
  try:
      print(a.next(wait=False))
  except Empty: pass
  try:
      print(b.next(wait=False))
  except Empty: pass

  # Wait for something to happen. The next time we loop either a or b should have data.
  conn.poll()
```

**Comment by deontologician** [comment]
We should definitely pick some underlying ioloop, rather than try to
implement our own. There are a ton of them (tornado, twisted, greenlet),
they're implemented in C, and they handle cross-platform decisions like
epoll/kqueue

**Comment by larkost** [comment]
@deontologician I think that we can get this job done without having a runloop, possibly without even having a separate thread. That being said, I do see an advantage for us to support one or more of the the async frameworks available for Python, and would prioritize them in this order: `twisted`, `greenlet`/`gevent`, then `tornado`. `twisted` first because it comes with Python on most distress, `greenlet`/`gevent` next because it is the most interesting and could give great demos, and finally `tornado` because it is a well known competitor in this space.

Does anyone have an objection to me separating the async aspect into its own issue?

**Comment by danielmewes** [comment]
@deontologician I agree that we should support those, but I still think that we should also provide a bare interface that allows concurrent changefeeds and doesn't rely on a specific async framework.
Note that the `poll` solution doesn't rely on asynchronous i/o. (Network) i/o can still be synchronous.

If I understand correctly, if we pick a single event loop / async io framework, our users will be forced to use that if they want to use our driver. So it sounds better to have a less convenient but framework-independent solution in the bare driver, and then provide customized drivers in addition to that.

**Comment by danielmewes** [comment]
@larkost No objections to opening separate threads for supporting specific async frameworks. We already have https://github.com/rethinkdb/rethinkdb/issues/2622 for Tornado.

**Comment by deontologician** [comment]
we talked about this in person and I agree it's a good idea to separate the async issue from the "can we do it" issue

**Comment by gchpaco** [comment]
> If I understand correctly, if we pick a single event loop / async io framework, our users will be forced to use that if they want to use our driver. So it sounds better to have a less convenient but framework-independent solution in the bare driver, and then provide customized drivers in addition to that.

This is actually why I was looking into lower level frameworks like the async stuff, and also for Ruby.  There are two or three different frameworks in this space (I think Trollop is probably worth supporting, for example) and they are not in any way compatible; making multiple event driven frameworks cooperate is an enormous job.  Better to support one standard way of doing it, and then add support for popular frameworks (like Twisted) as time permits.

**Comment by gchpaco** [comment]
FWIW, asyncio support exists for several of the common event loops (tulipcore and aiogevent for event, aiogreen and greenio for eventless, Tornado has support for it built in).  I don't see anything for Twisted, unfortunately.

**Comment by gchpaco** [comment]
It is also possible to do stuff with Trollius in Python <= 3.2 (3.3 can use the Tulip library and 3.4 has it built in) that is compatible with Tulip/asyncio, although it's a bit clumsy (since the workaround for the lack of `yield from` is of course not compatible.

**Comment by gchpaco** [comment]
Re: @coffeemug's proposal: if we permit `.on` to take a Python function, then we could make it a method on an object.  So this would be as follows:
```python
class Foo(object):
   # blah blah blah
f = Foo()
feed1 = r.table('foo').changes().run(conn).on(f.foo)
feed2 = r.table('bar').changes().run(conn).on(f.bar)
conn.poll(wait=True)
```
This would look (with perhaps some suitable glue code) like this in Tulip/asyncio:
```python
@asyncio.coroutine
def foo(change):
    while True:
        # some computation
        yield from asyncio.sleep(1)
@asyncio.coroutine
def bar(change):
    while True:
        # some computation
        yield from asyncio.sleep(1)
loop = asyncio_get_event_loop()
tasks = [
    r.table('foo').changes().run(conn).on(foo),
    r.table('bar').changes().run(conn).on(bar)
]
loop.run_until_complete(asyncio.wait(tasks))
loop.close()
```

**Comment by gchpaco** [comment]
Making this work with Trollius would look like this:
```python
import trollius

@trollius.coroutine
def foo(change):
    while True:
        # some computation
        yield trollius.From(asyncio.sleep(1))
@trollius.coroutine
def bar(change):
    while True:
        # some computation
        yield From(asyncio.sleep(1))
loop = trollius.get_event_loop()
tasks = [
    r.table('foo').changes().run(conn).on(foo),
    r.table('bar').changes().run(conn).on(bar)
]
loop.run_until_complete(trollius.wait(tasks))
loop.close()
```

**Comment by danielmewes** [comment]
@gchpaco Oh nice, so it looks like it would actually integrate with asyncio and Trollius very nicely.

**Comment by gchpaco** [comment]
Yeah, we would need to tag the `.on` results with one or the other decorator, which isn't that big a deal.

**Comment by gchpaco** [comment]
I looked into doing this with `gevent` and its event loop is very low level from what I saw.  I'm not entirely sure how to do it without getting into message queues and arcane details; I'm understanding better why Guido wrote Tulip.  And I've never understood Twisted very well.

**Comment by gchpaco** [comment]
Note that all of these would require hooking into whatever event system's event loop, of course.  So code might be slightly different, but not a lot different.

**Comment by gchpaco** [comment]
At a guess, `net.Connection` would need to have a parameter in the constructor specifying an `asyncio.BaseEventLoop` (may be better to have a separate subclass for this, actually) which would then use `asyncio.open_connection` instead of `socket.create_connection`; after that we would need to hook up `net.Cursor` to an event loop (again, probably best done as a separate subclass).  Um, `net.Connection._sock_recvall` and `net.Connection._sock_sendall` would need to be evented, which is probably mostly using the asyncio Stream interface.  There's a trick with writing on those that goes
```python
w.write(data)
yield from w.drain()
```
(or `yield From(w.drain())` in Trollius) but that's not a big deal.

Basically we would need to rebase our thing on top of `asyncio` which is deliberately not super difficult.

**Comment by danielmewes** [comment]
As far as the basic interface is concerned, we currently have two slightly different proposals:
1. Proposed by @larkost (+ `poll` added)
```Python
while True
  feedA = r.table('a').changes().run(conn)
  feedB = r.table('b').changes().run(conn)
  try:
      print(a.next(wait=False))
  except Empty: pass
  try:
      print(b.next(wait=False))
  except Empty: pass

  # Wait for something to happen. The next time we loop either feedA or feedB should have data.
  conn.poll()
```
2. Proposed by @coffeemug 
   ```Python
while True
  # Pass callbacks into on()
  feedA = r.table('a').changes().run(conn).on(lambda a: print(a))
  feedB = r.table('b').changes().run(conn).on(lambda b: print(b))

  # Wait for something to happen. This will call at least one of the callbacks
  conn.poll()
```

My impression from earlier comments in this thread is that callbacks are not very pythonic. However it might still be fine as a baseline interface.

Another question is whether one of the variants works better with event loop implementations. @gchpaco posted some sketches above that used @coffeemug's interface. @gchpaco since you've put some thought into this, do you think it would be similarly easy, easier or more difficult to do the same with the `cursor.next(wait=False)` interface?

**Comment by coffeemug** [comment]
I don't understand the first proposal. What happens when you call `poll`?

**Comment by gchpaco** [comment]
`cursor.next(wait=False)` is ... awkward in an event loop scenario, although it could be done.  The most awkward bit is that `conn.poll()` requires making some sort of decision as to which event loop we are using, because it amounts to `loop.run_forever()` if and only if the user is using an `asyncio` event loop.  If the user is using, say, gevent, then they need to get a `Greenlet` object and call `gevent.joinall(connections)`.

BTW the way I would put them is like this.  For `cursor.next(wait=False)`:
```python
feedA = r.table('a').changes().run(conn)
feedB = r.table('b').changes().run(conn)
while True:
  if feedA.hasData(): print(feedA.next(wait=False))
  if feedB.hasData(): print(feedB.next(wait=False))
  # Wait for something to happen. The next time we loop either feedA or feedB should have data.
  conn.poll()
```
And for the callback scenario:
```python
feedA = r.table('a').changes().run(conn).on(lambda a: print(a))
feedB = r.table('b').changes().run(conn).on(lambda b: print(b))

while True: conn.poll()
```
@larkost's proposal is very similar to a `select` loop in C, but clumsier because we are not returning the cursors that saw activity and so we must poll each one each time.  These are not easy to do for nontrivial amounts of work—for example in our simple demo here it has a bug wherein if `print` blocks (which it can because the other end is e.g. a pipe that is slow) responsiveness on the system goes to zero.  These "little problems" are *massively* irritating, and led in part to things like Node and its baroque callback system.

@coffeemug's solution at least has the benefit of using some kind of event loop situation—no polling loop, for one thing.  But we still have the problem where `conn.poll()` has to intuit whatever event system the user is proposing to use.  The proposals above appear to say "well, if the user doesn't specify an event loop then we should use our own (probably awful) hacked version".  Not 100% sure I like this, I think we would be better off picking something (probably Trollius because it offers an upgrade path in 3.3 and 3.4) and running with it, and then offering adapters for `gevent` and Tornado and whatever other thing crawls out of the woodwork.

**Comment by gchpaco** [comment]
> I don't understand the first proposal. What happens when you call poll?

It's just like `select` in C.

**Comment by danielmewes** [comment]
@coffeemug It just waits until some response is received from the server. So the next time you try `feedA.next(wait=False)` and `feedB.next(wait=False)` at least one of them will have data.

**Comment by danielmewes** [comment]
@gchpaco Ah yeah, I had the `while True` in the wrong place. Updated my post.

**Comment by coffeemug** [comment]
@gchpaco -- would you mind posting a minimal Python program that would consume data from two feeds via one of the standardized async APIs? (You seem to understand this quite well, and could probably do it 100x faster than I could research it)

I'm concerned that while many Python users will happily use a standardized event loop API, asking most users to use them to consume feeds might be too big a stretch. I think most users might just want to write a blocking command that returns feed events (e.g. `join(feedA, feedB)`). I'd like to look at a minimal async program to get a better idea if making that the only way to consume feeds is too big an ask.

**Comment by coffeemug** [comment]
To clarify, by `join` I mean something like `conn.get_next_blocking(list_of_feeds)`.

**Comment by gchpaco** [comment]
Certainly.  I'm going to presuppose some prior work has been done.

**Comment by gchpaco** [comment]
This isn't minimal, but it's sort of minimally interesting.
```python
import asyncio
import rethinkdb as r


queue = asyncio.Queue()


@asyncio.coroutine
def adder(change):
    current = 0
    while True:
        value = yield from change.next()
        current += value
        yield from queue.put(('adder', current))


@asyncio.coroutine
def multiplier(change):
    current = 1
    while True:
        value = yield from change.next()
        current *= value
        yield from queue.put(('multiplier', current))


@asyncio.coroutine
def combiner:
    best_adder = 0
    best_multiplier = 1
    while True:
        component, value = queue.get()
        if component == 'adder':
            best_adder = max(best_adder, value)
            yield (best_adder, best_multiplier)
        elif component == 'multiplier':
            best_multiplier = max(best_multiplier, value)
            yield (best_adder, best_multiplier)


@asyncio.coroutine
def printer:
    seen = 0
    for adder, multiplier in combiner():
        seen += 1
        print "Saw adder, multiplier pair %d, %d" % (adder, multiplier)
        if seen > 20:
            loop.stop()


loop = asyncio.get_event_loop()
asyncio.async(adder(r.table('a').count().changes().arun(conn)))
asyncio.async(multiplier(r.table('b').count().changes().arun(conn)))
loop.call_soon(printer, loop)
try:
    loop.run_forever()
finally:
    loop.close()
```
(minimally interesting because it actually processes things, etc).  This is more obnoxious in the callback case because of the need to maintain intermediate state.

If you just wanted to listen on two feeds it would look like this.
```python
@asyncio.coroutine
def a(change):
    while True:
        value = yield from change.next()
        pass


@asyncio.coroutine
def b(change):
    while True:
        value = yield from change.next()
        pass


loop = asyncio.get_event_loop()
asyncio.async(a(r.table('a').count().changes().arun(conn)))
asyncio.async(b(r.table('b').count().changes().arun(conn)))
loop.run_forever()
```

**Comment by gchpaco** [comment]
In person discussions with @coffeemug have revealed he would prefer an interface something like this:
```python
for index, change in r.join(r.table('a').count().changes(), r.table('b').count().changes()):
    if index == 0: # comes from table a
        pass
    elif index == 1: # comes from table b
        pass
```
which I think we can probably provide as porcelain for the user, that would use the `asyncio` stuff under the hood.  If the user wants to do something more sophisticated, they can of course use the `asyncio` machinery, but this may help with complete novices.

**Comment by coffeemug** [comment]
Talked to @gchpaco in person about this (a lot).

I agree that the `asyncio` API would be great for power users, but I think it's an overkill for basic tutorials, and would turn off a huge number of users. I think we should add a `join` command (possibly with a much better name) that would allow the user to wait on multiple cursors by returning a python generator:

```py
feedA = ...
feedB = ...
x = join(feedA, feedB)
for change in x:
   ...
```

Beginners (and even power users who are writing web apps) could use this API, and power users could use `asyncio` if they really need the features. @gchpaco also suggested implementing the `join` function on top of an `asyncio` event loop that would run entirely behind the scenes in our driver.

This solution makes sense to me, though I think there are a few more details to iron out in the API.

**Comment by coffeemug** [comment]
Err, @gchpaco beat me to it.

**Comment by gchpaco** [comment]
Ha!  My implementation plan, FWIW, would be to rebase the Python driver on top of `asyncio` with Trollius as a compatibility measure for Python < 3.3.  We would then add `loop=None` optional parameters probably to `r.run` and certainly to this proposed `r.join`, where the user could specify their own `asyncio` compatible event loop—adapters exist for gevent at least.

**Comment by danielmewes** [comment]
I don't think there's any need for changing our driver implementation to `asyncio` or Trollius for this interface.
In fact we can easily implement `r.join` using synchronous i/o only.

**Comment by gchpaco** [comment]
That'll only work if you only ever use one `r.join` in your entire program, though.  If you want to do anything nontrivial that interface gets clumsy.

**Comment by danielmewes** [comment]
`r.join` looks pretty nice though for simple use cases.

One disadvantage is that all changefeeds must be known in advance, and you cannot easily add another changefeed to the join later (though I guess we could add a method on the object `r.join` returns to join something else or to remove a changefeed from it). The callback-based interface is easier to use in that respect.
For simple applications that might be sufficient, and if we later provide an `asyncio` interface for more advanced use cases I can imagine that this would satisfy most users.

**Comment by danielmewes** [comment]
I don't see why multiple `r.join` would be a problem.
The connection would simply keep track of all open cursors and cursors would have a method that allows the connection to push a result to them. I think we already do this in the current implementation.
`r.join(...).next()` would just make the connection keep reading from the network until it has added at least one result to one of the cursors that this `r.join()` is watching.

**Comment by gchpaco** [comment]
Yeah, it's porcelain, for people to get bootstrapped with.  If we do it with some event loop under the hood then they can even have multiple `r.join`s concurrently—not really possible with synchronous I/O.  But I don't want to tart it up so it's suitable for every power user.  We could provide an `asyncio` compatible interface for them.

FWIW it's not really easy to just hook our driver into an event loop; they tend to want to use their own socket methods.

Your proposed implementation amounts to implementing a miniature `select` loop, which I suppose we could do, but it's still awkward.

**Comment by coffeemug** [comment]
I think not to implementing `asyncio` for 2.0/v1, and just adding `r.join` (with a better name) to allow reading from multiple feeds/cursors is a great solution. It would probably take care of 99% of early use cases (if not more), and we can always add full blown `asyncio` support later if demand arises.

If we wanted to get fancy, we could even do this:

```py
feedA = ...
feedB = ...
x = r.join(feedA, feedB)

feedC = ...
feedD = ...
y = r.join(feedC, feedD)

z = r.join(x, y)
for change in z:
  ...
```

We could also do the following as @danielmewes suggested:

```py
feedE = ...
z.add(feedE)
z.remove(feedA)
```

This seems like a very powerful API to me. I would stick with it and ignore `asyncio` for the first implementation, unless there is a *really* compelling reason not to.

**Comment by coffeemug** [comment]
Man, that actually sounds really awesome. I'd personally *love* to program in this thing.

**Comment by danielmewes** [comment]
> I think not to implementing asyncio for 2.0/v1, and just adding r.join (with a better name) to allow reading from multiple feeds/cursors is a great solution.

:+1: We can (and should) ship extended drivers that use certain async frameworks lazily after 2.0.


One thing to keep in mind for the "joining joins" or "adding/removing from a join object" interface is that we need to assign useful indexes to newly added member-feeds. I think this might become a bit annoying to use with the `x = r.join(...); y = r.join(...); z = r.join(x, y)` interface, but we could make
```
z.add(...)
```
return the index of the newly added feed, and also maintain the invariant that removing a feed from a joint feed through
```
z.remove(...)
```
keeps the indexes of all other member feeds intact.

It's also worth pointing out that the `r.join` interface can be mapped almost 1:1 onto server-side union changefeeds, e.g.
```
r.union(
    feedA.map(r.row.merge({feedIndex: 0})),
    feedB.map(r.row.merge({feedIndex: 1}))
  ).changes()
```

However server-side `union.changes` won't allow to add/remove feeds to the union dynamically, so I think it's still worth adding the client-side `r.join` (or whichever name we pick) interface. Another advantage is that it provides us with a test-bed for porting our driver to async frameworks, even if we don't do it in 2.0.

**Comment by danielmewes** [comment]
@larkost you actually need multiple changefeeds for the Python-based tests. Would the `r.join` interface as described by @gchpaco https://github.com/rethinkdb/rethinkdb/issues/3298#issuecomment-73007642 work for your use case?

**Comment by timmaxw** [comment]
I don't think `r.join` is a good idea. It forces users to manually disentangle updates from multiple changefeeds, which gets ugly fast. I think callbacks are the most natural way to consume changefeeds.

The API I'd like to see for 2.0 is the `.run().on()` / `.poll()` one. I think it's about as powerful as `r.join()` and much nicer to program in. So the "Hello World" example would be:
```
import rethinkdb as r
conn = r.connect()
def callback(change):
    print change
r.table("foo").changes().run().on(callback)
conn.loop_forever()
```
I don't think this will confuse or scare away new users.

I agree that changefeeds are ugly and non-Pythonic if people want to chain a long series of *non-changefeed* queries asynchronously. In the short term, I don't think we should support asynchronous non-changefeed queries. In the long term, I think we should support `asyncio`. Our basic example on the front page of rethinkdb.com would look like this under `asyncio`:

```python
import rethinkdb as r
import asyncio
@asyncio.coroutine
def main():
    conn = yield from r.connect(event_loop = asyncio.get_event_loop())
    yield from r.db("test").table_create("tv_shows").arun(conn)
    yield from r.table("tv_shows").insert([
        { name: 'Star Trek TNG', episodes: 178 },
        { name: 'Battlestar Galactica', episodes: 75 }
        ]).run(conn)
    print(yield from r.table("foo").count().arun(conn))
    iterator = yield from r.table("foo").filter(r.row["episodes"].gt(100)).arun(conn)
    try:
        while True:
            print(yield from iterator.next())
    except StopIteration:
        pass
loop = asyncio.get_event_loop()
loop.run_until_complete(main)
loop.close()
```

**Comment by gchpaco** [comment]
Note that @timmaxw's example is more complex than necessary because it uses all this async machinery to do a simple linear series of updates.  It's more compelling when you want to e.g. do sixteen inserts concurrently and wait for them all to finish.

**Comment by gchpaco** [comment]
If we rebase the Python driver onto `asyncio` then `run`, `connect` and probably cursors would become something like this:
```
def run(*args, *kwargs):
    loop = asyncio.get_event_loop()
    result = arun(*args, *kwargs)
    loop.run_until_complete(result)
    return result.result()
```
and similarly for the others.  That is, they will automatically play nice with async stuff under the scenes, so you don't need to use `arun` and `aconnect` and the like unless actually necessary.

**Comment by timmaxw** [comment]
If we implement `query.run()` as @gchpaco suggested, we wouldn't even need a `conn.loop_forever()` method. Instead, it would just be:
```python
import rethinkdb as r
conn = r.connect()
def callback(change):
    print change
r.table("foo").changes().run().on(callback)
import asyncio
asyncio.get_event_loop().run_forever()
```

**Comment by deontologician** [comment]
Count me onboard for the callbacks for changefeeds option. I'm concerned about making the basic "just get started" version significantly different than the code you'd need to write real applications in. It seems the `r.join` syntax does force the user to write a select loop which is annoying and handled by all of these async libraries already.

Ultimately, the users and the driver will be gravitating towards asyncio over time, so it makes sense to align the driver with that api vs. inventing something specific for our needs. A lot of time and thought went into the asyncio interface, and we probably shouldn't waste time coming up with something unique.

**Comment by gchpaco** [comment]
I would make `conn.loop_forever()` exist, but be a direct call into `run_forever()` on the associated event loop.  Easier to explain to new users, does the same thing.

**Comment by coffeemug** [comment]
> ```js
r.union(
    feedA.map(r.row.merge({feedIndex: 0})),
    feedB.map(r.row.merge({feedIndex: 1}))
  ).changes()
```

Friendly reminder -- you can run `merge` on streams (and feeds) today. So the above code could be written like this:

 ```js
r.union(
    feedA.merge({feedIndex: 0}),
    feedB.merge({feedIndex: 1})
  ).changes()
```

Or better yet:

 ```js
r.union(
    feedA.merge({source: 'feedA'}),
    feedB.merge({source: 'feedB'})
  ).changes()
```

WRT to the API:

- it sounds like we're in agreement that we shouldn't bother with `asyncio` in v1.
- I'd like to spend 10m with @mglukhovsky tomorrow before we proceed. He tends to have really good feedback for things like these.

**Comment by danielmewes** [comment]
In an offline discussion with @larkost , @coffeemug , @timmaxw and @mlucy , we decided to provide an event-loop integration already in 2.0.

Apart from https://github.com/rethinkdb/rethinkdb/issues/3529 , this is going to be the only way to read from multiple changefeeds at the same time in Python.

We will be providing an `on(cb)` function on cursors, and integrate with the framework's event loop to call the callback once we get a response.

An open question was which frameworks we should support first.

Tornado seems to be very widely used and easily available, so I think we should definitely support that.

As a second one @gchpaco suggested asyncio for Python >= 3.4 and the similar Trollius for older versions of Python. It probably makes sense to support that second.

**Comment by danielmewes** [comment]
@gchpaco said he would start looking into a Tornado integration.

**Comment by danielmewes** [comment]
Edit: Sorry I had written Tornado in my two previous comments, but actually meant Twisted. Updated my posts.

**Comment by danielmewes** [comment]
Marking as settled. We can discuss details in follow-up issues for specific async / event libraries.

**Comment by danielmewes** [comment]
Closing this.

It will be solved by either of
https://github.com/rethinkdb/rethinkdb/issues/2622
https://github.com/rethinkdb/rethinkdb/issues/3529
https://github.com/rethinkdb/rethinkdb/issues/3785
(and further ports we might do in the future)
